### PR TITLE
Bigquery sync table require filter

### PR DESCRIPTION
**Pull request by qnkhuat**
First milestone of https://github.com/metabase/metabase/issues/36668

Known limitation: this won't work for materialized views, see https://github.com/metabase/metabase/pull/36677/files#diff-370f1464c7c10bd0d0bd84544e6f80102152617fcd99fea6e9bd4dbaff4903caR131-R135 for why

To read more about partitioned tables, see https://cloud.google.com/bigquery/docs/partitioned-tables

### CODE REVIEW

File: src/metabase/sync/sync_metadata/tables.clj [hunk 1]
```diff
@@ -189,32 +209,32 @@
   ([database :- i/DatabaseInstance db-metadata]
    ;; determine what's changed between what info we have and what's in the DB
    (let [db-tables               (table-set db-metadata)
-         our-metadata            (our-metadata database)
-         strip-desc              (fn [metadata]
-                                   (set (map #(dissoc % :description) metadata)))
+         name+schema->db-table   (m/index-by (juxt :name :schema) db-tables)
```
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/tables.clj:212]: I took a stab at refactoring this sync step a bit.
Details in the following comments

File: src/metabase/sync/sync_metadata/tables.clj [hunk 2]
```diff
@@ -189,32 +209,32 @@
...
      (when (some? (:version db-metadata))
        (sync-util/with-error-handling (format "Error creating/reactivating tables for %s"
                                               (sync-util/name-for-logging database))
          (update-database-metadata! database db-metadata)))
      ;; create new tables as needed or mark them as active again
      (when (seq new-tables)
-       (sync-util/with-error-handling (format "Error creating/reactivating tables for %s"
-                                              (sync-util/name-for-logging database))
-         (create-or-reactivate-tables! database new-tables)))
+       (let [new-tables-info (set (map #(get name+schema->db-table [(:name %) (:schema %)]) new-tables))]
```
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/tables.clj:226]: previously, when creating a table, we didn't include a table description; we relied on the update step to do that. that's necessary, so I've changed it so that the description is included in the creation step.

File: src/metabase/sync/sync_metadata/tables.clj [hunk 3]
```diff
@@ -189,32 +209,32 @@
...
      ;; mark old tables as inactive
      (when (seq old-tables)
        (sync-util/with-error-handling (format "Error retiring tables for %s" (sync-util/name-for-logging database))
          (retire-tables! database old-tables)))
 
-     ;; update description for changed tables
-     (when (seq changed-tables)
-       (sync-util/with-error-handling (format "Error updating table description for %s" (sync-util/name-for-logging database))
-         (update-table-description! database changed-tables)))
+     (sync-util/with-error-handling (format "Error updating table metadata for %s" (sync-util/name-for-logging database))
```
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/tables.clj:235]: previously, only table description could be updated during sync step, but I've changed it so that it could extend to other properties too.

File: src/metabase/sync/sync_metadata/tables.clj [hunk 4]
```diff
@@ -140,27 +142,47 @@
...
+   metabase-table :- (ms/InstanceOf :model/Table)]
+  (log/infof "Updating table metadata for %s" (sync-util/name-for-logging metabase-table))
+  (let [to-update-keys [:description :database_require_filter]
+        old-table      (select-keys metabase-table to-update-keys)
+        new-table      (select-keys (merge
+                                     (zipmap to-update-keys (repeat nil))
+                                     table-metadata)
+                                    to-update-keys)
+        [_ changes _]  (data/diff old-table new-table)
+        changes        (cond-> changes
```
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/tables.clj:164]: this is a specical case, check test `metabase.sync.sync-metadata.comments-test/sync-existing-table-comment-test` and `metabase.sync.sync-metadata.comments-test/dont-overwrite-table-custom-description-test` to know how it's expected to work.

File: src/metabase/sync/sync_metadata/tables.clj [hunk 5]
```diff
@@ -189,32 +209,32 @@
   ([database :- i/DatabaseInstance db-metadata]
    ;; determine what's changed between what info we have and what's in the DB
    (let [db-tables               (table-set db-metadata)
-         our-metadata            (our-metadata database)
-         strip-desc              (fn [metadata]
-                                   (set (map #(dissoc % :description) metadata)))
+         name+schema->db-table   (m/index-by (juxt :name :schema) db-tables)
+         our-metadata            (db->our-metadata database)
+         keep-table-id-info      (fn [metadata]
```
Comment (qnkhuat) [inline src/metabase/sync/sync_metadata/tables.clj:214]: This change somehow changes the order of sample dataset tables added on startup. 

that's why you'll see some tests affected by this, one of which is I have to update the table ides in `e2e/support/cypress_data.js`

File: test/metabase/sync/sync_metadata/comments_test.clj [hunk 6]
```diff
@@ -114,7 +117,14 @@
       (mt/dataset (basic-table "table_with_comment_after_sync" nil)
         ;; modify the source DB to add the comment and resync
         (driver/notify-database-updated driver/*driver* (mt/db))
-        (tx/create-db! driver/*driver* (basic-table "table_with_comment_after_sync" "added comment"))
+        (tx/create-db! driver/*driver* (basic-table "table_with_comment_after_sync" nil))
```
Comment (qnkhuat) [inline test/metabase/sync/sync_metadata/comments_test.clj:120]: This test relied on the fact that descriptions are not included when inserting tables, the correct behavior is to start with a table without a description, then sync, then add a description, then sync. see https://github.com/metabase/metabase/pull/36677/files#r1427842665 for context.

File: test/metabase/automagic_dashboards/core_test.clj [hunk 7]
```diff
@@ -185,8 +185,8 @@
   (mt/with-test-user :rasta
     (automagic-dashboards.test/with-dashboard-cleanup
       (doseq [[table cardinality] (map vector
-                                       (t2/select Table :db_id (mt/id) {:order-by [[:id :asc]]})
-                                       [15 11 7 5])]
+                                       (t2/select Table :db_id (mt/id) {:order-by [[:name :asc]]})
```
Comment (qnkhuat) [inline test/metabase/automagic_dashboards/core_test.clj:188]: There are 2 things happening here:
- this is a consequence of test tables ids being changed ([context](https://github.com/metabase/metabase/pull/36677/files#r1427855029)).
- after https://github.com/metabase/metabase/pull/35973, the test-data database now has 8 tables. so that's why you see my changes also include 4 more items.

File: e2e/support/cypress_data.js [hunk 8]
```diff
@@ -25,15 +25,15 @@ export const SAMPLE_DB_SCHEMA_ID = "1:PUBLIC";
 // Use only for e2e helpers and custom commands. Never in e2e tests directly!
```
Comment (qnkhuat) [inline e2e/support/cypress_data.js:25]: see https://github.com/metabase/metabase/pull/36677/files#r1427855029 for why I have this change

File: src/metabase/sync/sync_metadata/tables.clj [hunk 9]
```diff
@@ -140,27 +142,47 @@
...
+        old-table      (select-keys metabase-table to-update-keys)
+        new-table      (select-keys (merge
+                                     (zipmap to-update-keys (repeat nil))
+                                     table-metadata)
+                                    to-update-keys)
+        [_ changes _]  (data/diff old-table new-table)
+        changes        (cond-> changes
+                         ;; we only update the description if and only if the initial state is nil
+                         ;; because don't want to override the user edited description if it exists
+                         (some? (:description old-table))
```
Comment (calherries) [inline src/metabase/sync/sync_metadata/tables.clj:167]: Before, the check was `(str/blank? ...)` not `(some? ...)`. Why did you change it?

**Review by calherries — CHANGES_REQUESTED** [review]

File: src/metabase/sync/sync_metadata/tables.clj (hunk 5 above)
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/tables.clj:214]: it's a function so I think the name should contains a verb

how about `keep-schema+name-set`?

File: src/metabase/sync/sync_metadata/tables.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/tables.clj:226]: we can't fetch the value from the diff because the diff only contains `schema + name`, whereas `create-or-reactivate-tables` need to get the full info from `db-tables`.

File: src/metabase/sync/sync_metadata/tables.clj (hunk 9 above)
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/tables.clj:167]: I think that was a mistake. If a table does not have a description, then the jdbc will return `NULL`.

whereas if a table has a description and it's empty, then it might be a case where the user manually removes the description in Metabase.

In this case, we don't want to override that on update.

File: test/metabase/sync/sync_metadata/comments_test.clj [hunk 10]
```diff
@@ -114,7 +117,14 @@
...
+        ;; create the comment
+        (jdbc/execute! (sql-jdbc.conn/db->pooled-connection-spec (mt/db))
+                       [(sql.tx/standalone-table-comment-sql
+                         driver/*driver*
+                         {:name "table_with_comment_after_sync_db"}
+                         {:table-name "table_with_comment_after_sync"
+                          :table-comment "added comment"})])
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/sync/sync_metadata/comments_test.clj]: addresed in cb4955805e

File: test/metabase/sync/sync_metadata/tables_test.clj [hunk 11]
```diff
@@ -39,3 +43,20 @@
...
+(deftest sync-table-update-info-of-new-table-added-during-sync-test
+  (testing "during sync, if a table is reactived, make sure we do updates the table info if needed"
+    (mt/dataset (mt/dataset-definition "sync-retired-table"
+                  ["user" [{:field-name "name" :base-type :type/Text}] [["Ngoc"]]])
+      (t2/update! :model/Table (mt/id :user) {:active false})
+      ;; table description is changed
+      (jdbc/execute! (sql-jdbc.conn/db->pooled-connection-spec (mt/db))
+                     [(sql.tx/standalone-table-comment-sql
+                       (:engine (mt/db))
+                       {:name "sync-retired-table"}
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/sync/sync_metadata/tables_test.clj]: cb4955805e

File: src/metabase/sync/sync_metadata/tables.clj [hunk 12]
```diff
@@ -189,32 +209,32 @@
...
-                                              (sync-util/name-for-logging database))
-         (create-or-reactivate-tables! database new-tables)))
+       (let [new-tables-info (set (map #(get name+schema->db-table [(:name %) (:schema %)]) new-tables))]
+         (sync-util/with-error-handling (format "Error creating/reactivating tables for %s"
+                                                (sync-util/name-for-logging database))
+           (create-or-reactivate-tables! database new-tables-info))))
      ;; mark old tables as inactive
      (when (seq old-tables)
        (sync-util/with-error-handling (format "Error retiring tables for %s" (sync-util/name-for-logging database))
          (retire-tables! database old-tables)))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/tables.clj:236]: yea you're right. this sync code is quite old, so I think a lot of things need to be improved here.

anw I applied your suggestion here cb4955805e

File: src/metabase/sync/sync_metadata/tables.clj [hunk 13]
```diff
@@ -140,27 +142,47 @@
...
+    (when (seq changes)
+      (t2/update! :model/Table (:id metabase-table)  changes))))
+
+(mu/defn ^:private update-tables-metadata-if-needed!
+  [table-metadatas :- [:set i/DatabaseMetadataTable]
+   metabase-tables :- [:set (ms/InstanceOf :model/Table)]]
+  (let [name+schema->table-metadata (m/index-by (juxt :name :schema) table-metadatas)
+        name+schema->metabase-table (m/index-by (juxt :name :schema) metabase-tables)]
+    (doseq [[name+schema metabase-table] name+schema->metabase-table]
+      (when-let [table-metadata (name+schema->table-metadata name+schema)]
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/tables.clj]: cb4955805e

File: e2e/test/scenarios/admin/datamodel/hide_tables.cy.spec.js [hunk 14]
```diff
@@ -16,15 +16,15 @@ describe("scenarios > admin > datamodel > hidden tables (metabase#9759)", () =>
 
   it("hidden table should not show up in various places in UI", () => {
     // Visit the main page, we shouldn't be able to see the table
-    cy.visit(`/browse/${PRODUCTS_ID}`);
+    cy.visit(`/browse/${SAMPLE_DB_ID}`);
```
Comment (qnkhuat) [inline e2e/test/scenarios/admin/datamodel/hide_tables.cy.spec.js:19]: this was a mistake and it accidentally works because `producs_id` and `sample_db_id` had the same id (=1)

File: src/metabase/sync/sync_metadata/tables.clj (hunk 2 above)
Comment (calherries) [reply src/metabase/sync/sync_metadata/tables.clj:229]: Yes, that's why I passed the diff output through `(->> <diff> (map schema+name->table))`

Sorry for the poor formatting I was taking shortcuts to write this in the github UI.

Here's the correct formatting:
```
[new-tables old-tables] (->> (data/diff (set (map schema+name db-tables)) (set (map schema+name our-metadata)))
                             (map schema+name->table))
```
Of course you could do this in two steps. My main point is that `schema+name` can be extracted out into a function and used to do the diff and getting the full info.

File: src/metabase/sync/sync_metadata/tables.clj (hunk 5 above)
Comment (calherries) [reply src/metabase/sync/sync_metadata/tables.clj:217]: Function names don't need to contain verbs, but sure `keep-schema+name-set` is better too.

File: src/metabase/sync/sync_metadata/tables.clj (hunk 9 above)
Comment (calherries) [reply src/metabase/sync/sync_metadata/tables.clj:171]: Makes sense!

File: modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj [hunk 15]
```diff
@@ -240,6 +249,60 @@
                    {:source-table (mt/id view-name)
                     :order-by     [[:asc (mt/id view-name :id)]]}))))))))
 
+(deftest sync-table-with-required-filter-test
```
Comment (calherries) [inline modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj:252]: I think you should add a test for materialized views too, based on this line 
https://github.com/metabase/metabase/pull/36677/files#diff-370f1464c7c10bd0d0bd84544e6f80102152617fcd99fea6e9bd4dbaff4903caR130

File: src/metabase/sync/sync_metadata/tables.clj [hunk 16]
```diff
@@ -136,31 +140,52 @@
...
+  (log/infof "Updating table metadata for %s" (sync-util/name-for-logging metabase-table))
+  (let [to-update-keys [:description :database_require_filter]
+        old-table      (select-keys metabase-table to-update-keys)
+        new-table      (select-keys (merge
+                                     (zipmap to-update-keys (repeat nil))
+                                     table-metadata)
+                                    to-update-keys)
+        [_ changes _]  (data/diff old-table new-table)
+        changes        (cond-> changes
+                         ;; we only update the description if and only if the initial state is nil
```
Comment (calherries) [inline src/metabase/sync/sync_metadata/tables.clj:169]: redundant "only"
```suggestion
                         ;; we only update the description if the initial state is nil
```

File: modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj (hunk 15 above)
Comment (qnkhuat) [reply modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj:252]: right so turns out we currently can't figure out whether a materialized view have required a filter or not. See https://github.com/metabase/metabase/pull/36677/files#diff-370f1464c7c10bd0d0bd84544e6f80102152617fcd99fea6e9bd4dbaff4903caR131-R135 for comments.

I'll add this as a limitation in the PR

File: src/metabase/sync/sync_metadata/tables.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/sync/sync_metadata/tables.clj]: did that here https://github.com/metabase/metabase/pull/36677/commits/ce46551884f37acbd888e40e00d1d377d738df2b#diff-8a4b5a6ee44180ba51e170fe36de9d5f3f889cd95b1ee56d19c72b6aac130ae5R215-R222

File: modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj (hunk 15 above)
Comment (calherries) [reply modules/drivers/bigquery-cloud-sdk/test/metabase/driver/bigquery_cloud_sdk_test.clj:252]: Makes sense, thanks for investigating

File: modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj [hunk 17]
```diff
@@ -123,6 +115,34 @@
...
+(defmethod driver/describe-database :bigquery-cloud-sdk
+  [_ database]
+  (let [tables (list-tables (:details database))]
+    {:tables (set (for [^Table table tables
+                        :let  [^TableId         table-id   (.getTableId table)
+                               ^String          dataset-id (.getDataset table-id)
+                               ^TableDefinition tabledef   (.getDefinition table)
+                               table-name                  (str (.getTable table-id))]]
+                    {:schema                  dataset-id
+                     :name                    #pp table-name
```
Comment (calherries) [inline modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj:127]: You left a #pp here

**Review by calherries — APPROVED** [review]
Solid work

**Review by calherries — CHANGES_REQUESTED** [review]
Actually sorry, I think you're missing a test case for a table that is initially created without a required filter, then synced, then is modified to have a required filter, then synced again. Similar to `sync-existing-table-comment-test`

File: e2e/test/scenarios/admin/datamodel/hide_tables.cy.spec.js (hunk 14 above)
Comment (nemanjaglumac) [reply e2e/test/scenarios/admin/datamodel/hide_tables.cy.spec.js:19]: 🤦‍♂️

**Comment by qnkhuat** [comment]
@calherries 

>  I think you're missing a test case for a table that is initially created without a required filter, then synced, then is modified to have a required filter, then synced again. Similar to sync-existing-table-comment-test

This is not possible in bigquery because whether a table requires partition or not is a property it's created with. afaik You can't change it after.

I could add sth like `sync-table-update-info-of-new-table-added-during-sync-test` tho.

**Comment by calherries** [comment]
Based on the docs ([here](https://cloud.google.com/bigquery/docs/samples/bigquery-update-table-require-partition-filter#bigquery_update_table_require_partition_filter-java) and [here](https://www.javadoc.io/doc/com.google.apis/google-api-services-bigquery/v2-rev20200719-1.30.10/com/google/api/services/bigquery/model/Table.html#setRequirePartitionFilter-java.lang.Boolean-)), it looks like you can update the property after the table is created.

**Comment by calherries** [comment]
Also using SQL ([source](https://stackoverflow.com/questions/59636355/bigquery-how-to-set-require-partition-filter-by-tablereference)):
```
ALTER TABLE IF EXISTS mydataset.newtable
SET OPTIONS(
    require_partition_filter = true
)
```

**Comment by qnkhuat** [comment]
Good find! I was looking at the list of options for `ALTER TABLE SET OPTIONS` command([docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#column_set_options_list)) and didn't find this option so I thought it's not possible.

anw I added the test here 8452b0b2a2

**Review by calherries — APPROVED** [review]
Great, thanks!
